File: java/150.evaluate-reverse-polish-notation.java
/*
 * @lc app=leetcode id=150 lang=java
 *
 * [150] Evaluate Reverse Polish Notation
 */

// @lc code=start
class Solution {
    public int evalRPN(String[] tokens) {
        Deque<Integer> stack = new ArrayDeque<>();

        int op2, op1;

        for(String s : tokens) {
            switch(s) {
                case "+":
                    op2 = stack.pop();
                    op1 = stack.pop();
                    stack.push(op1 + op2);
                    break;
                case "-":
                    op2 = stack.pop();
                    op1 = stack.pop();
                    stack.push(op1 - op2);
                    break;
                case "*":
                    op2 = stack.pop();
                    op1 = stack.pop();
                    stack.push(op1 * op2);
                    break;
                case "/":
                    op2 = stack.pop();
                    op1 = stack.pop();
                    stack.push(op1 / op2);
                    break;
                default:
                    stack.push(Integer.parseInt(s));
            }
        }        

        return stack.pop();
    }
}
// @lc code=end

File: cpp/150.evaluate-reverse-polish-notation.cpp
/*
 * @lc app=leetcode id=150 lang=cpp
 *
 * [150] Evaluate Reverse Polish Notation
 */

// @lc code=start
class Solution {
public:
    int evalRPN(vector<string>& tokens) {

        std::stack<int> stk;

        std::unordered_map<std::string, std::function<void()>> map = {
            {"+", [&](){
                int y = stk.top();
                stk.pop();
                int x = stk.top();
                stk.pop();
                stk.push(x + y);
            }},
            {"-", [&](){
                int y = stk.top();
                stk.pop();
                int x = stk.top();
                stk.pop();
                stk.push(x - y);
            }},
            {"*", [&](){
                int y = stk.top();
                stk.pop();
                int x = stk.top();
                stk.pop();
                stk.push(x * y);
            }},
            {"/", [&](){
                int y = stk.top();
                stk.pop();
                int x = stk.top();
                stk.pop();
                stk.push(x / y);
            }}
        }; 

        for(std::string& s : tokens) {
            if(map.contains(s)) {
                map[s]();
            }
            else {
                stk.push(std::stoi(s));
            }
        }

        return stk.top();
    }
};
// @lc code=end

